### PR TITLE
feat(frontend): define `ConstantEvalRewriter`, impl `ExprRewritable` for batch

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Next steps:
- impl for stream nodes

Depends on: https://github.com/risingwavelabs/risingwave/pull/7542

Let's remember to enable `now.slt.part` after we enable `ConstantEvalRewriter`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/7538

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/expr.yaml [hunk 1]
```diff
@@ -144,20 +144,19 @@
     -- Single quoted literal can be treated as number without error.
     values(round('123'));
   batch_plan: |
-    BatchValues { rows: [[Round(123:Float64)]] }
+    BatchValues { rows: [[123:Float64]] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/expr.yaml:147]: maybe let's move const_eval into separate PR so it's easier to review

File: src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs [hunk 2]
```diff
@@ -136,12 +132,11 @@ impl_plan_tree_node_for_binary! { StreamDynamicFilter }
 impl StreamNode for StreamDynamicFilter {
     fn to_stream_prost_body(&self, state: &mut BuildFragmentGraphState) -> NodeBody {
         use generic::dynamic_filter::*;
-        let condition = self.core.predicate().as_expr_unless_true().map(|x| {
-            self.base
-                .ctx()
-                .expr_with_session_timezone(x)
-                .to_expr_proto()
-        });
+        let condition = self
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:135]: We are not considering the dynamic filter condition an expr as it has to be of a rigid form i.e. not rewritable.

File: src/frontend/planner_test/tests/testdata/expr.yaml (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/planner_test/tests/testdata/expr.yaml:147]: This PR will be for const eval, the exprrewritable part is now here: https://github.com/risingwavelabs/risingwave/pull/7542

File: src/frontend/src/optimizer/plan_node/batch_filter.rs [hunk 3]
```diff
@@ -93,3 +95,22 @@ impl ToLocalBatch for BatchFilter {
...
+
+    fn rewrite_exprs(&self, r: &mut dyn ExprRewriter) -> PlanRef {
+        Self {
+            base: self.base.clone_with_new_plan_id(),
+            logical: self
+                .logical
+                .rewrite_exprs(r)
+                .as_logical_filter()
+                .unwrap()
+                .clone(),
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/batch_filter.rs:112]: Hopefully this will be optimized away by the compiler...

**Comment by jon-chuang** [comment]
Oops, we can't enable the optimization yet. The session timezone substitution has to happen before or we end up with a cast error.

**Comment by st1page** [comment]
> Oops, we can't enable the optimization yet. The session timezone substitution has to happen before or we end up with a cast error. Disable for now.

I think it can be a common optimization and not only for `now()` expression in batch :thinking:

**Comment by jon-chuang** [comment]
> Disable for now.

I meant disable for the time being XD

**Comment by chenzl25** [comment]
Since we have already solved the `Share` node clone problem, can we use the existing `new` of `create` method of each plan node and avoid constructing the plan node by `Self` directly? In this way we can remove the `clone_with_new_plan_id`  interface as well. Rest LGTM.

File: src/frontend/src/optimizer/plan_node/expr_rewritable.rs [hunk 4]
```diff
@@ -18,31 +18,40 @@ use paste::paste;
 
 use super::*;
 use crate::expr::ExprRewriter;
-use crate::{for_batch_plan_nodes, for_stream_plan_nodes};
+use crate::for_stream_plan_nodes;
 
 /// Rewrites expressions in a `PlanRef`. Due to `Share` operator,
 /// the `ExprRewriter` needs to be idempotent i.e. applying it more than once
 /// to the same `ExprImpl` will be a noop on subsequent applications.
 pub trait ExprRewritable {
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/expr_rewritable.rs:26]: Add rustdocs to clarify that it only rewrite exprs in a plan, and link to `RewriteExprsRecursive`? 🤔

File: src/frontend/src/optimizer/plan_node/expr_rewritable.rs [hunk 5]
```diff
@@ -18,31 +18,40 @@ use paste::paste;
...
-    fn rewrite_exprs(&self, r: &mut dyn ExprRewriter) -> PlanRef;
+    fn has_rewritable_expr(&self) -> bool {
+        false
+    }
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/expr_rewritable.rs:29]: Why is this needed? (compared with a no-op `rewrite_exprs` impl)

File: src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    pub(crate) error: Option<RwError>,
+}
+impl ExprRewriter for ConstEvalRewriter {
+    fn rewrite_expr(&mut self, expr: ExprImpl) -> ExprImpl {
+        if self.error.is_some() {
+            return expr;
+        }
+        if expr.is_const() {
+            let data_type = expr.return_type();
+            match expr.eval_row_const() {
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs:29]: I'm still a little bit worried about the implementation of `eval_row_const`. But maybe it's not a big deal and the interface doesn't have any problem 🥵

File: src/frontend/src/optimizer/plan_node/expr_rewritable.rs (hunk 5 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/expr_rewritable.rs:29]: It's just so we don't have to clone the plan.

File: src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs (hunk 6 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs:29]: Yes, there is overhead to it... But at least we evaluate it just once, whereas if we propagate non-const-folded exprs to e.g. stream jobs, then we will incur linear overhead.

**Comment by jon-chuang** [comment]
> can we use the existing new or create method of each plan node and avoid constructing the plan node by Self directly?

I think this makes sense for most batch nodes, but for logical nodes, I think its still more straigtforward and maintainable to instantiate from Self.

**Comment by chenzl25** [comment]
> > can we use the existing new or create method of each plan node and avoid constructing the plan node by Self directly?
> 
> I think this makes sense for most batch nodes, but for logical nodes, I think its still more straigtforward and maintainable to instantiate from Self.

+1, make sense.

**Review by chenzl25 — APPROVED** [review]
LGTM
